### PR TITLE
fix: revolution number computation

### PR DESCRIPTION
**Pull request by apaletta3**
Fixes the off-by-one bug in the `PropagatedModel.calculateRevolutionNumberAt()` method.

Bug: The break check uses strict inequalities (>, <), but it is performed before revolutionNumber += durationSign and before currentInstant += durationSign * orbitalPeriod for the current iteration. As a result, when the loop reaches a currentInstant that is exactly equal to requesting instant, the predicate is still false, so the code increments the revolution counter one extra time and only breaks on the next iteration after overshooting by a full orbital period.
Result: 
- When asking for a rev number before epoch, you would get the true rev number minus one
- When asking for a rev number before epoch, you would get the true rev number plus one

New logic:
- Start from the epoch state
- compute the orbital period, jump back or forwards by one revolution
- break with the current rev number of we are past the requested instant
- otherwise increment/decrement the rev number

Note:
- Additionally, I also avoid repropagating all the way from the epoch state to the currentState each time the loop runs, because that is computationally slower than just propagating from the previously propagated state. 
- I also generalized the logic in the Kepler model to use the `getRevlutionNumberAtEpoch()` instead of assuming it is always 1, even though it is hardcoded to that

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed revolution-number origin and epoch handling so orbital revolution counts are consistent across models and propagation methods, including edge cases near node crossings.
  * Updated propagation logic to compute periods from the current orbit state for more reliable stepping.

* **Tests**
  * Expanded and refactored tests to use period-based validation, cover edge cases, and assert valid integer results.

### CODE REVIEW

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Kepler.cpp [hunk 1]
```diff
@@ -233,17 +233,20 @@ Integer Kepler::calculateRevolutionNumberAt(const Instant& anInstant) const
     switch (perturbationType_)
     {
         case Kepler::PerturbationType::None:
-            return Kepler::CalculateNoneRevolutionNumberAt(coe_, epoch_, gravitationalParameter_, anInstant);
+            return Kepler::CalculateNoneRevolutionNumberAt(coe_, epoch_, gravitationalParameter_, anInstant) +
```
Comment (apaletta3) [inline src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Kepler.cpp:236]: `Kepler::CalculateNoneRevolutionNumberAt` is a static method and so can't access `this->getRevolutionNumberAtEpoch()`, hence why I add the value here

Also `Kepler::CalculateNoneRevolutionNumberAt` is a private method, so behavior doesn't change for the user here

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp [hunk 2]
```diff
@@ -319,16 +310,6 @@ Integer Propagated::calculateRevolutionNumberAt(const Instant& anInstant) const
...
-
-        // Propagate for duration of this orbital period
-        const State currentState = propagator_.calculateStateAt(
-            cachedStateArray_[0], cachedStateArray_[0].accessInstant() + (durationSign * orbitalPeriod)
-        );
-
-        // Update the current instant position and velocity coordinates
-        currentPositionCoordinates = currentState.getPosition().getCoordinates();
-        currentVelocityCoordinates = currentState.getVelocity().getCoordinates();
-        currentInstant += durationSign * orbitalPeriod;
```
Comment (apaletta3) [inline src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp:331]: the issue with the previous logic was that `currentInstant` wasn't updated before the `revolutionNumber` was incremented/decremented

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp [hunk 3]
```diff
@@ -272,44 +275,32 @@ Integer Propagated::calculateRevolutionNumberAt(const Instant& anInstant) const
...
     // Determine whether to count revolution numbers in forwards or backwards time and return function if duration is 0
-    Instant currentInstant = cachedStateArray_[0].getInstant();
-    const double durationInSecs = (anInstant - currentInstant).inSeconds();
-    if (durationInSecs == 0.0)
-    {
-        return 1;
-    }
-    Integer durationSign = (durationInSecs > 0) - (durationInSecs < 0);
+    const State stateAtEpoch = this->calculateStateAt(this->getEpoch());
+    const Integer durationSign = (anInstant > this->getEpoch()) ? 1 : -1;
```
Comment (alex-liang3) [inline src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp:300]: nit: why not use the `Sign` type?

**Review by alex-liang3 — APPROVED** [review]

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp [hunk 4]
```diff
@@ -272,44 +275,32 @@ Integer Propagated::calculateRevolutionNumberAt(const Instant& anInstant) const
...
-            -gravitationalParameter_SI * currentPositionCoordinates.norm() /
-            (currentPositionCoordinates.norm() * std::pow(currentVelocityCoordinates.norm(), 2) -
-             2.0 * gravitationalParameter_SI);
-        const Duration orbitalPeriod =
-            Duration::Seconds(Real::TwoPi() * std::sqrt(std::pow(semiMajorAxis, 3) / gravitationalParameter_SI));
+        // Calculate orbital period at current state
+        const COE coe = COE::Cartesian(
+            {currentState.getPosition(), currentState.getVelocity()}, Earth::Spherical.gravitationalParameter_
+        );  // (Spherical earth has the most modern value which is the correct one)
+        const Duration orbitalPeriod = coe.getOrbitalPeriod(Earth::Spherical.gravitationalParameter_);
```
Comment (vishwa2710) [inline src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp]:
```suggestion
        TBI: allow user to provide gravitational parameter
        const Duration orbitalPeriod = coe.getOrbitalPeriod(Earth::Spherical.gravitationalParameter_);
```

**Review by vishwa2710 — APPROVED** [review]

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp (hunk 3 above)
Comment (apaletta3) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp:300]: not sure I'm able to increment smoothly with it

File: src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp (hunk 4 above)
Comment (apaletta3) [reply src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit/Model/Propagated.cpp]: I'll do one better - get the grav param from the propagator_
